lib/ecto/repo/preloader.ex

```diff
@@ -152,8 +153,40 @@ defmodule Ecto.Repo.Preloader do
     [entry | preload_assocs(assocs, queries, repo_name, tuplet)]
   end
 
+
   defp preload_assocs([], [], _repo_name, _tuplet), do: []
 
+  defp preload_embeds(structs, [], _repo_name, _tuplet), do: structs
+
+  defp preload_embeds(structs, [embed | embeds], repo_name, tuplet) do
+    {%{field: field, cardinality: card}, sub_preloads} = embed
+    {embed_structs, counts} =
+      Enum.flat_map_reduce(structs, [], fn struct, counts ->
+        case {Map.get(struct, field), card} do
+          {nil, _card} ->{[], [0 | counts]}
+          {embed_struct, :one} -> {[embed_struct], [1 | counts]}
+          {embed_structs, :many} -> {embed_structs, [length(embed_structs) | counts]}
+        end
+      end)
+    embed_structs = preload_each(embed_structs, repo_name, sub_preloads, tuplet)
```

Comment:
```suggestion
    {%{field: field, cardinality: card}, sub_preloads} = embed
    
    {embed_structs, counts} = 
      Enum.flat_map_reduce(structs, [], fn 
        %{^field => embeds}, counts when is_list(embeds) -> {embeds, [length(embeds) | counts]}
        %{^field => embed}, counts -> {[embed], [1 | counts]}
        nil, counts -> {[], [0 | counts]}
        struct, _counts -> raise ArgumentError, "expected #{inspect(struct)} to contain embed `#{field}`"
      end)
      
    embed_structs = preload_each(embed_structs, repo_name, sub_preloads, tuplet)
```

@Sleepful sorry last change then we can merge :). You discovered that `structs` can have `nil` inside of it so we need to consider it. Also just some minor formatting (adding new lines between single line statements and multiline statements).

Reply:
> You discovered that structs can have nil inside of it so we need to consider it

Oh no, I am not sure about this. Current code already handles the case when `embeds_one` is `nil` and I just added tests where `embeds_many` is `nil` instead of `[]`, which I am not sure if it is possible but it might be so why not handle that case too.

But is this last change necessary?

Reply:
If you can think of a test case that fails without this suggestion but passes with the suggestion, that would be great for me to understand.

Reply:
oh sure so in your integration test instead of massing `[order1, order2, order3, order4]` into preload pass `[order1, nil, order2, order3, order4]`.

It will try to do `Map.get(nil, field)` and raise an error

Reply:
Ditto! Review the updated test case